sdks/python/apache_beam/io/filesystem.py

```diff
@@ -531,24 +530,117 @@ def _list(self, dir_or_prefix):
     """
     raise NotImplementedError
 
+  @staticmethod
+  def _split_scheme(url_or_path):
+    match = re.match(r'(^[a-z]+)://(.*)', url_or_path)
+    if match is not None:
+      return match.groups()
+    return None, url_or_path
+
+  @staticmethod
+  def _combine_scheme(scheme, path):
+    if scheme is None:
+      return path
+    return '{}://{}'.format(scheme, path)
+
   def _url_dirname(self, url_or_path):
     """Like posixpath.dirname, but preserves scheme:// prefix.
 
     Args:
       url_or_path: A string in the form of scheme://some/path OR /some/path.
     """
-    match = re.match(r'([a-z]+://)(.*)', url_or_path)
-    if match is None:
-      return posixpath.dirname(url_or_path)
-    url_prefix, path = match.groups()
-    return url_prefix + posixpath.dirname(path)
+    scheme, path = self._split_scheme(url_or_path)
```

Comment:
At some point while trying out `PurePath` I used the schema splitting in `match_files` as well. I felt that the methods could stay since there may be other uses for them.

Reply:
Although that might be a case of unnecessary indirection.

Reply:
I'm not sure on this one. I do think the splitting does make it more readable.